psi4/src/psi4/fisapt/fisapt.cc

```diff
@@ -2230,6 +2230,441 @@ void FISAPT::disp(std::map<std::string, SharedMatrix> matrix_cache, std::map<std
+        dfh->fill_tensor("Cas", Cas, {astart, astart + nablock});
+        dfh->fill_tensor("Das", Das, {astart, astart + nablock});
+        dfh->fill_tensor("Ear", Ear, {astart, astart + nablock});
+
+        for (size_t bstart = 0; bstart < nb; bstart += max_b) {
+            size_t nbblock = (bstart + max_b >= nb ? nb - bstart : max_b);
+
+            dfh->fill_tensor("Abs", Abs, {bstart, bstart + nbblock});
+            dfh->fill_tensor("Bbs", Bbs, {bstart, bstart + nbblock});
+            dfh->fill_tensor("Cbr", Cbr, {bstart, bstart + nbblock});
+            dfh->fill_tensor("Dbr", Dbr, {bstart, bstart + nbblock});
+            dfh->fill_tensor("Ebs", Ebs, {bstart, bstart + nbblock});
+
+            long int nab = nablock * nbblock;
+
+#pragma omp parallel for schedule(dynamic) reduction(+ : Disp20, CompleteDisp20)
+            for (long int ab = 0L; ab < nab; ab++) {
+                int a = ab / nbblock;
+                int b = ab % nbblock;
+
+                int thread = 0;
+#ifdef _OPENMP
+                thread = omp_get_thread_num();
+#endif
+
+                double** Trsp = Trs[thread]->pointer();
+                double** Vrsp = Vrs[thread]->pointer();
+
+                // => Amplitudes, Disp20 <= //
+
+                C_DGEMM('N', 'T', nr, ns, nQ, 1.0, Aarp[(a)*nr], nQ, Absp[(b)*ns], nQ, 0.0, Vrsp[0], ns);
+
+                for (int r = 0; r < nr; r++) {
+                    for (int s = 0; s < ns; s++) {
+                        Trsp[r][s] = Vrsp[r][s] / (eap[a + astart] + ebp[b + bstart] - erp[r] - esp[s]);
+                        Disp20 += 4.0 * Trsp[r][s] * Vrsp[r][s];
+                    }
+                }
+
+                // => Exch-Disp20 <= //
+
+                // > DF-Part < //
+
+                C_DGEMM('N', 'T', nr, ns, nQ, 4.0, Barp[(a)*nr], nQ, Bbsp[(b)*ns], nQ, 0.0, Vrsp[0], ns);
+                C_DGEMM('N', 'T', nr, ns, nQ, -2.0, Cbrp[(b)*nr], nQ, Casp[(a)*ns], nQ, 1.0, Vrsp[0], ns);
+                C_DGEMM('N', 'T', nr, ns, nQ, -2.0, Dbrp[(b)*nr], nQ, Dasp[(a)*ns], nQ, 1.0, Vrsp[0], ns);
+                C_DGEMM('N', 'T', nr, ns, nQ, 4.0, Earp[(a)*nr], nQ, Ebsp[(b)*ns], nQ, 1.0, Vrsp[0], ns);
+
+                // > AO-Part < //
+
+                C_DGER(nr, ns, 1.0, Tarp[a + astart], 1, omega_bsp[b + bstart], 1, Vrsp[0], ns);
+                C_DGER(nr, ns, -1.0, omega_brp[b + bstart], 1, Tasp[a + astart], 1, Vrsp[0], ns);
+                C_DGER(nr, ns, 1.0, omega_arp[a + astart], 1, Tbsp[b + bstart], 1, Vrsp[0], ns);
+                C_DGER(nr, ns, -1.0, Tbrp[b + bstart], 1, omega_asp[a + astart], 1, Vrsp[0], ns);
+                for (int r = 0; r < nr; r++) {
+                    for (int s = 0; s < ns; s++) {
+                        CompleteDisp20 += Trsp[r][s] * Vrsp[r][s];
+                    }
+                }
+            }
+        }
+    }
+
+    double ExchDisp20 = CompleteDisp20 - Disp20;
+
+    scalars_["Disp20"] = Disp20;
+    scalars_["Exch-Disp20 (S^inf)"] = ExchDisp20;
```

Comment:
for testing, you've collected the Sinf Exch-Disp20 here. Add a line analogous to https://github.com/psi4/psi4/blob/master/psi4/src/psi4/fisapt/fisapt.cc#L2334 . After that, the output of your new function will be available in any input file as `variable('SAPT EXCH-DISP20(S^2) ENERGY')`.

You probably have this value lying around for a few systems from either your reference implementation or from putting the term into psi4. Go ahead and put those molecules in a test file, run the `energy(...)` as you have been, then add a line `compare_values(your_ref_value, variable('sapt exch-disp20(s^2) energy'), 5, msg)` where the 5 (or whatever's appropriate is the number of decimal places to compare.

Reply:
After adding the line, I'm getting the following error when I run my water.input:
Traceback (most recent call last):
  File "/home/relik107/psi4/objdir/stage/bin/psi4", line 287, in <module>
    exec(content)
  File "<string>", line 47, in <module>
  File "/home/relik107/psi4/objdir/stage/lib/psi4/driver/p4util/python_helpers.py", line 573, in _core_variable
    raise KeyError("psi4.core.variable: Requested variable " + key + " was not set!\n")

KeyError: 'psi4.core.variable: Requested variable sapt exch-disp20(s^inf) energy was not set!\n'

Reply:
what's your `git diff`?

Reply:
diff --git a/psi4/driver/procrouting/sapt/sapt_mp2_terms.py b/psi4/driver/procrouting/sapt/sapt_mp2_terms.py
index 6c83cc4c4..795ba19c0 100644
--- a/psi4/driver/procrouting/sapt/sapt_mp2_terms.py
+++ b/psi4/driver/procrouting/sapt/sapt_mp2_terms.py
@@ -243,10 +243,6 @@ def df_mp2_fisapt_dispersion(wfn, primary, auxiliary, cache, do_print=True):
     fisapt.sinf_disp(matrix_cache, vector_cache, False)
     scalars = fisapt.scalars()
 
-    core.print_out("\n")
-    core.print_out(print_sapt_var("Disp20 (MP2)", scalars["Disp20"], short=True) + "\n")
-    core.print_out(print_sapt_var("Exch-Disp20,u (S^inf)", scalars["Exch-Disp20 (S^inf)"], short=True) + "\n")
-
     return ret
 
 
diff --git a/psi4/src/psi4/fisapt/fisapt.cc b/psi4/src/psi4/fisapt/fisapt.cc
index 11f31240f..7158cb18a 100644
--- a/psi4/src/psi4/fisapt/fisapt.cc
+++ b/psi4/src/psi4/fisapt/fisapt.cc
@@ -2659,6 +2659,7 @@ void FISAPT::sinf_disp(std::map<std::string, SharedMatrix> matrix_cache, std::ma
 
     scalars_["Disp20"] = Disp20;
     scalars_["Exch-Disp20 (S^inf)"] = ExchDisp20;
+    Process::environment.globals["SAPT EXCH-DISP20(S^inf) ENERGY"] = scalars_["Exch-Disp20 (S^inf)"];
     if (do_print) {
         outfile->Printf("    Disp20              = %18.12lf [Eh]\n", Disp20);
         outfile->Printf("    Exch-Disp20 (S^inf) = %18.12lf [Eh]\n", ExchDisp20);
diff --git a/tests/sapt-exch-disp-inf/water.input b/tests/sapt-exch-disp-inf/water.input
index fb9242ac8..7f046bb32 100644
--- a/tests/sapt-exch-disp-inf/water.input
+++ b/tests/sapt-exch-disp-inf/water.input
@@ -1,4 +1,4 @@
-memory 62 gb
+memory 4 gb
  
 molecule dimer {
 0 1
@@ -30,3 +30,5 @@ set globals {
 
 set SAPT_DFT_FUNCTIONAL HF
 energy('sapt(dft)', molecule=dimer)
+
+compare_values(4.35491, variable('sapt exch-disp20(s^inf) energy'), 5, "SAPT EXCH-DISP20(S^inf) ENERGY")

Reply:
try capitalizing that `INF` in your added line. psi is less case-tolerant on the programmer's side. Good catch that I meant "^inf" but said "^2" in previous post.